xds/internal/client/callback.go

```diff
@@ -74,26 +74,50 @@ func (c *clientImpl) callCallback(wiu *watcherInfoWithUpdate) {
 //
 // A response can contain multiple resources. They will be parsed and put in a
 // map from resource name to the resource content.
-func (c *clientImpl) NewListeners(updates map[string]ListenerUpdate) {
+func (c *clientImpl) NewListeners(updates map[string]ListenerUpdate, metadata UpdateMetadata) {
 	c.mu.Lock()
 	defer c.mu.Unlock()
 
+	if metadata.ErrState != nil {
+		// On NACK, update overall version to the NACKed resp.
+		c.ldsVersion = metadata.ErrState.Version
+		for name := range updates {
+			if _, ok := c.ldsWatchers[name]; ok {
```

Comment:
Just confirming my understanding:
With the new way of doing things, let's say we initially had listeners `[A, B]`. Then we get an update where we decide that the newly received `B` is something that we don't like. So, we will update the metadata associated with the resources, and will NACK the update, but will continue using the old `[A, B]`?
Is that right?

Reply:
Right. This doesn't change how updates are processed. It only add metadata.

There's a TODO to send the NACK error to the watcher, then it's up to the watcher to ignore the error (and keep using the old resource) or fail RPCs.
E.g. the xds_resolver will ignore errors and keep using old resources, but will fail RPCs if the first resp is a NACK